common/src/main/java/com/microsoft/identity/common/adal/internal/util/JsonExtensions.java

```diff
@@ -0,0 +1,31 @@
+package com.microsoft.identity.common.adal.internal.util;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+
+/**
+ * Helper methods related to JSON
+ */
+public class JsonExtensions {
+
+    public static Map<String, String> extractJsonObjectIntoMap(final String jsonString)
```

Comment:
I know this is just being 'pulled-in' from ADAL but there's something about it's functionality I thought I would at least call your attention to: it 'flattens' JSON objects, meaning nested objects need to call this method multiple times to access inner-fields.

For instance:
```json
{
	"a": "Aardvark",
	"b": "Banana",
	"c": "Cardamom",
	"d": {
		"1": "Cyclops",
		"2": "Biclops",
		"3": "Triclops",
		"4": "Quadclops?"
	}
}
```

Returns a `Map<String, String` containing four entries:
```
"a" : "Aardvark"
"b" : "Banana"
"c" : "Cardamom"
"d" : "{"1":"Cyclops","2":"Biclops","3":"Triclops","4":"Quadclops?"}"
```

I went digging around in the code to test this before I posted this comment - I also found _three_ implementations of it peppered around the code:

1. [IdToken](https://github.com/AzureAD/azure-activedirectory-library-for-android/blob/dev/adal/src/main/java/com/microsoft/aad/adal/IdToken.java#L161-L170)
2. [OAuth2](https://github.com/AzureAD/azure-activedirectory-library-for-android/blob/dev/adal/src/main/java/com/microsoft/aad/adal/Oauth2.java#L342-L352)
3. [SignInActivity](https://github.com/AzureAD/azure-activedirectory-library-for-android/blob/dev/automationtestapp/src/main/java/com/microsoft/aad/automation/testapp/SignInActivity.java#L208-L217)

Reply:
Yep.  Saw that.  In this case it's ok... I'll file another issue around improving JSON handling in general.  This method avoids deserializing to a specific type... which makes sense in some cases and not in others.  You could argue that this isn't the correct method for dealing with ClientInfo for example...